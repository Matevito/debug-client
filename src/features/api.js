const axios = require('axios');

const api = axios.create({
    baseURL: "https://pure-falls-26749.herokuapp.com/apiv1/",
    timeout: 3000
});

export default api;