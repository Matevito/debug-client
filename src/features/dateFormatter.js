import { DateTime } from "luxon"

const simpleDate = (string) => {
    return DateTime.fromISO(string).toLocaleString(DateTime.DATE_MED)
};

const detailedDate = (string) => {
    return DateTime.fromISO(string).toLocaleString(DateTime.DATETIME_FULL_WITH_SECONDS)
}

export {
    simpleDate,
    detailedDate
}