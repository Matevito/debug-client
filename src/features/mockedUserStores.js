import configureStore from "redux-mock-store";

// todo: set project an issues values
const usersStoreValues = [
    {
        user: {} 
    },
    {
        user: {
            loggedIn: true,
            token : "a token?",
            user: {
                username: "devUser",
                email:"dev@email.com",
                role: "Developer",
                id: "1"
            },
            projects: {
                number: 0,
                list: []
            },
            issues: {
                number: 0,
                list: []
            }
        }
    },
    {

        user: {
            loggedIn: true,
            token : "a token?",
            user: {
                username: "TeamLeaderUser",
                email:"team@email.com",
                role: "Team leader",
                id: "2"
            },
            projects: {
                number: 0,
                list: []
            },
            issues: {
                number: 0,
                list: []
            }
        }
    },
    {

        user: {
            loggedIn: true,
            token : "admintoken?",
            user: {
                username: "adminUser",
                email:"admin@email.com",
                role: "Admin",
                id: "3"
            },
            projects: {
                number: 0,
                list: []
            },
            issues: {
                number: 0,
                list: []
            }
        }
    }
];

exports.getData = () => {
    const mockedUsers = usersStoreValues.map((user, i) => {
        const mockStore = configureStore([]);
        if (i === 0) {
            return mockStore(user)
        }
        return mockStore({user: {
            user: {
                user
            }
        }})
    });

    return mockedUsers;
}