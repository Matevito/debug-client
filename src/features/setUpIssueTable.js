const setUpIssueTable = (issuesList) => {
    let data = [
        {label: "open", x: 1,  y: 0},
        {label: "aditional info needed", x: 2, y: 0},
        {label: "in progress", x: 3, y: 0},
        {label: "under review", x: 4, y: 0},
        {label: "solved",x: 5, y: 0}
    ];

    issuesList.forEach((issue) => {
        data.forEach((obj, index) => {
            if (obj.label === issue.status){
                data[index].y += 1
            }
        })
    });
    
    return data.filter(obj => obj.y > 0);
};


export default setUpIssueTable;