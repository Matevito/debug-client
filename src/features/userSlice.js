import  { createSlice } from "@reduxjs/toolkit";

/* User object structure
user:{
    loggedIn: 'boolean',
    user: {
        username,
        email,
        role,
        id
    },
    projects: [{
        id,
        title
    }],
    issues: [{
        id,
        title
    }],
    todo on it's own slice:?
    notifications: [array of id's] // under development*
}
*/
export const userSlice = createSlice({
    name: 'user',
    initialState: {
        user: null
    },
    reducers: {
        login: (state, action) => {
            state.user = action.payload
        },
        logout: (state) => {
            state.user = null
        }
    }
});

export const { login, logout } = userSlice.actions;

export const selectUser = (state) => state.user.user;

export default userSlice.reducer