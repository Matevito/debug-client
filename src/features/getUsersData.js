import api from "./api";

const getUserData = async (token) => {
    const config = {
        headers: { "auth-token" : token}
    }; 
    try {
        const apiRes = await api.get("/user/list", config);
        return apiRes.data.data
    } catch(err) {
        return null
    }
}

export default getUserData