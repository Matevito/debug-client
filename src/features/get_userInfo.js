import api from "./api";

const get_userInfo = async (token) => {
    // setting up headers for calls
    const config = {
        headers: {
            "auth-token": token
        }
    };
    // fetch api data
    const userData = await api.get("/whoami", config);
    if (!userData) {
        return false;
    }
    const userId = userData.data.data.id;
    const userRelatedData = await api.get(`/user/${userId}`, config);
    if (!userRelatedData) {
        return false;
    };

    const userNotifications = [];

    // structure res and send it
    const dataRes = {
        loggedIn: true,
        token: token,
        user: {
            username: userData.data.data.username,
            email: userData.data.data.email,
            role: userData.data.data.role,
            id: userData.data.data.id
        },
        projects: userRelatedData.data.data.projects,
        issues: userRelatedData.data.data.issues,
        notifications: {
            list: userNotifications,
            number: userNotifications.length
        }
    };
    return dataRes;
}

export default get_userInfo;