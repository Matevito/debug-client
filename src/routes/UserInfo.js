import React from 'react'
import { useNavigate } from "react-router-dom";

// redux comp
import { useSelector } from "react-redux";
import { selectUser } from "../features/userSlice";



export const UserInfo = () => {
    const user = useSelector(selectUser);
    let navigate = useNavigate();

    if (!user) {
        navigate("/")
    } else {
        return (
            <>todo...</>
        )
    }
}
