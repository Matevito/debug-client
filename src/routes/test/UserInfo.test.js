import { Provider } from "react-redux";
import { MemoryRouter, Route, Routes } from "react-router-dom";

// testing comps
import { render, screen } from "@testing-library/react";
import '@testing-library/jest-dom'

//test component
import { UserInfo } from "../UserInfo";

// mock functs
import { getData } from "../../features/mockedUserStores";

const mockedUsedNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedUsedNavigate,
}));

const renderComponent = (url, store) => {
    return render(
        <Provider store={store}>
            <MemoryRouter initialEntries={[url]}>
                <Routes>
                    <Route 
                        path="/user/info"
                        element={<UserInfo />}
                    />
                </Routes>
            </MemoryRouter>
        </Provider>
    )
}
describe("UserInfo component", () => {
    let noLoggedUser;
    let devStore;
    beforeEach(() => {
        const usersData = getData();
        noLoggedUser = usersData[0];
        devStore = usersData[1];
    });
    afterAll(() => {
        jest.clearAllMocks();
    })
    test("handles no logged user", () => {
        renderComponent("/user/info", noLoggedUser)
        expect(mockedUsedNavigate).toHaveBeenCalledWith("/")
});
    test("renders user info card", () => {
        const { container } = renderComponent("/user/info", devStore);
        expect(container).toMatchSnapshot();
    });
})