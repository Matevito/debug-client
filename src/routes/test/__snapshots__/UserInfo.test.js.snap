// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserInfo component renders user info card 1`] = `
<div>
  <div
    class="MuiBox-root css-af1f4c"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation8 MuiCard-root css-1r7yq5s-MuiPaper-root-MuiCard-root"
    >
      <div
        class="MuiCardContent-root css-1vjpr1i-MuiCardContent-root"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="AccountBoxIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 5v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2H5c-1.11 0-2 .9-2 2zm12 4c0 1.66-1.34 3-3 3s-3-1.34-3-3 1.34-3 3-3 3 1.34 3 3zm-9 8c0-2 4-3.1 6-3.1s6 1.1 6 3.1v1H6v-1z"
          />
        </svg>
        <h4
          class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
        >
          DEVUSER
        </h4>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 css-2bmocg-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1hn59hz-MuiGrid-root"
          >
            <div
              class="css-1kq5d21-MuiStack-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="BadgeIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 7h-5V4c0-1.1-.9-2-2-2h-2c-1.1 0-2 .9-2 2v3H4c-1.1 0-2 .9-2 2v11c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V9c0-1.1-.9-2-2-2zM9 12c.83 0 1.5.67 1.5 1.5S9.83 15 9 15s-1.5-.67-1.5-1.5S8.17 12 9 12zm3 6H6v-.75c0-1 2-1.5 3-1.5s3 .5 3 1.5V18zm1-9h-2V4h2v5zm5 7.5h-4V15h4v1.5zm0-3h-4V12h4v1.5z"
                />
              </svg>
              <h6
                class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
              >
                <b>
                  User role
                </b>
              </h6>
            </div>
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              Developer
            </p>
            <div
              class="css-1kq5d21-MuiStack-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="EmailIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 4-8 5-8-5V6l8 5 8-5v2z"
                />
              </svg>
              <h6
                class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
              >
                <b>
                  Contact email
                </b>
              </h6>
            </div>
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              dev@email.com
            </p>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1hn59hz-MuiGrid-root"
          >
            <div
              class="css-147rfyl-MuiStack-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="GroupsIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 12.75c1.63 0 3.07.39 4.24.9 1.08.48 1.76 1.56 1.76 2.73V18H6v-1.61c0-1.18.68-2.26 1.76-2.73 1.17-.52 2.61-.91 4.24-.91zM4 13c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm1.13 1.1c-.37-.06-.74-.1-1.13-.1-.99 0-1.93.21-2.78.58C.48 14.9 0 15.62 0 16.43V18h4.5v-1.61c0-.83.23-1.61.63-2.29zM20 13c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm4 3.43c0-.81-.48-1.53-1.22-1.85-.85-.37-1.79-.58-2.78-.58-.39 0-.76.04-1.13.1.4.68.63 1.46.63 2.29V18H24v-1.57zM12 6c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3z"
                />
              </svg>
              <h6
                class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
              >
                <b>
                  Projects user is part 
                </b>
              </h6>
              <h5
                class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
              >
                2
              </h5>
            </div>
            <div
              class="css-147rfyl-MuiStack-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="BugReportIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 8h-2.81c-.45-.78-1.07-1.45-1.82-1.96L17 4.41 15.59 3l-2.17 2.17C12.96 5.06 12.49 5 12 5c-.49 0-.96.06-1.41.17L8.41 3 7 4.41l1.62 1.63C7.88 6.55 7.26 7.22 6.81 8H4v2h2.09c-.05.33-.09.66-.09 1v1H4v2h2v1c0 .34.04.67.09 1H4v2h2.81c1.04 1.79 2.97 3 5.19 3s4.15-1.21 5.19-3H20v-2h-2.09c.05-.33.09-.66.09-1v-1h2v-2h-2v-1c0-.34-.04-.67-.09-1H20V8zm-6 8h-4v-2h4v2zm0-4h-4v-2h4v2z"
                />
              </svg>
              <h6
                class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
              >
                <b>
                  Tickets taken
                </b>
              </h6>
              <h5
                class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
              >
                0
              </h5>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
